core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,268 @@
+        "As a result reporting of ProcessTree metrics is stopped")
+        isAvailable = false
+        return -1
+    }
+  }
+
+  private def computePageSize(): Unit = {
+    val cmd = Array("getconf", "PAGESIZE")
+    val out: Array[Byte] = Array.fill[Byte](10)(0)
+    Runtime.getRuntime.exec(cmd).getInputStream.read(out)
+    pageSize = Integer.parseInt(new String(out, "UTF-8").trim)
+  }
+
+  private def computeProcessTree(): Unit = {
+    if (!isAvailable) {
+      return
+    }
+    computePageSize
+    val queue: Queue[Int] = new Queue[Int]()
+    queue += pid
+    while( !queue.isEmpty ) {
+      val p = queue.dequeue()
+      val c = getChildPIds(p)
+      if(!c.isEmpty) {
+        queue ++= c
+        ptree += (p -> c.toSet)
+      }
+      else {
+        ptree += (p -> Set[Int]())
+      }
+    }
+  }
+
+  private def getChildPIds(pid: Int): ArrayBuffer[Int] = {
+    try {
+      val cmd = Array("pgrep", "-P", pid.toString)
+      val input = Runtime.getRuntime.exec(cmd).getInputStream
+      val childPidsInByte: mutable.ArrayBuffer[Byte] = new mutable.ArrayBuffer()
+      var d = input.read()
+      while (d != -1) {
+        childPidsInByte.append(d.asInstanceOf[Byte])
+        d = input.read()
+      }
+      input.close()
+      val childPids = new String(childPidsInByte.toArray, "UTF-8").split("\n")
+      val childPidsInInt: ArrayBuffer[Int] = new ArrayBuffer[Int]()
+      for (p <- childPids) {
+        if (p != "") {
+          childPidsInInt += Integer.parseInt(p)
+        }
+      }
+      childPidsInInt
+    } catch {
+      case e: IOException => logDebug("IO Exception when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped")
+        isAvailable = false
+        return new mutable.ArrayBuffer()
+      case _ => logDebug("Some exception occurred when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped")
+        isAvailable = false
+        return new mutable.ArrayBuffer()
+    }
+  }
+
+  /**
+   * Hadoop ProcfsBasedProcessTree class used regex and pattern matching to retrive the memory
+   * info. I tried that but found it not correct during tests, so I used normal string analysis
+   * instead. The computation of RSS and Vmem are based on proc(5):
+   * http://man7.org/linux/man-pages/man5/proc.5.html
+   */
+  def getProcessInfo(pid: Int): Unit = {
+    try {
+      val pidDir: File = new File(procfsDir, pid.toString)
+      val fReader = new InputStreamReader(
+        new FileInputStream(
+          new File(pidDir, procfsStatFile)), Charset.forName("UTF-8"))
+      val in: BufferedReader = new BufferedReader(fReader)
+      val procInfo = in.readLine
+      in.close
+      fReader.close
+      val procInfoSplit = procInfo.split(" ")
+      if ( procInfoSplit != null ) {
+        if (procInfoSplit(1).toLowerCase.contains("java")) {
```

Comment:
the checks for java & python seem error prone (maybe you have some external program called "pass-data-from-java-to-fizzbuzz-cpp") ... but I'm also not sure what else you could do here.

Reply:
So I'm checking for the child processes of a Spark process. And there you will see either java or pythonX.Y. I don't think that Spark will create a process with custom name

Reply:
It might not be as uncommon to have arbitrary forked processes from Spark executors as we think. Recent discussions from Spark Summit 2018 to my knowledge propose sending off work to be done by specialized hardware or specialized software. I'm not an expert in these use cases so someone can correct this observation if they like. Mostly just flagging that for completeness.